baseplate/frameworks/queue_consumer/kafka.py

```diff
@@ -0,0 +1,506 @@
+            # Note: It is recommended to set enable.auto.offset.store=false for
+            # long-time processing applications and then explicitly store offsets
+            # after message processing, to make sure offsets are not auto-committed
+            # prior to processing has finished.
+            "max.poll.interval.ms": 300000,
+            # disable offset autocommit, we'll manually commit.
+            "enable.auto.commit": "false",
+        }
+
+    def build_message_handler(self) -> KafkaMessageHandler:
+        assert self.message_handler_count == 0, "Can only run 1 message handler!"
+
+        self.message_handler_count += 1
+
+        # pylint: disable=unused-argument
+        def commit_offset(
+            context: RequestContext, data: Dict[str, Any], message: confluent_kafka.Message
+        ) -> None:
+            logger.debug(
+                "committing topic %s partition %s offset %s",
+                message.topic(),
+                message.partition(),
+                message.offset(),
+            )
+            with context.trace.make_child("kafka.commit"):
+                self.consumer.commit(message=message, asynchronous=False)
+
+        return KafkaMessageHandler(
+            self.baseplate,
+            self.name,
+            self.handler_fn,
+            self.message_unpack_fn,
+            # commit offset after each successful message handle()
+            on_success_fn=commit_offset,
+        )
+
+
+class FastConsumerFactory(_BaseKafkaQueueConsumerFactory):
+    """Factory for running a :py:class:`~baseplate.server.queue_consumer.QueueConsumerServer` using Kafka.
+
+    The `FastConsumerFactory` prioritizes high throughput over exactly once
+    message processing.
+
+    This will run a single `KafkaConsumerWorker` that reads messages from Kafka and
+    puts them into an internal work queue. Then it will run multiple `KafkaMessageHandler`s
+    that read messages from the internal work queue, processes them with the
+    `handler_fn`. The number of `KafkaMessageHandler` processes is controlled
+    by the `max_concurrency` parameter in the `~baseplate.server.queue_consumer.QueueConsumerServer`
+    configuration. Kafka partition offsets are automatically committed by the
+    `confluent_kafka.Consumer` every 5 seconds, so any message that has been
+    read by the `KafkaConsumerWorker` could be committed, regardless of whether
+    it has been processed.
+
+    This server should be able to achieve very high message processing throughput
+    due to the multiple `KafkaMessageHandler` processes and less frequent, background
+    partition offset commits. This does come at a price though: messages may be
+    processed out of order, not at all, or multiple times. This is appropriate
+    when processing throughput is important and it's acceptable to skip messages
+    or process messages more than once (maybe there is ratelimiting in the
+    handler or somewhere downstream).
```

Comment:
🔕 🔕 🔕 

Potential idea for the future, build a basic "idempotence" checker that you can pass in a `backend` (redis or memcached) and a callback to get the `key` to use to dedupe (could even to default to something set automatically by a future baseplate kafka producer client) that can run a basic idempotence check to guard against processing a message more than once so we don't have to re-implement that any time we want to use `FastConsumerFactory`.

Reply:
🔕